### PR TITLE
feat: implement blocking web hooks

### LINKED ISSUE #1724: Setting up a Criteria for self-service registration
The feature is meant to set custom criteria for the registration form, for example setting up a dob (date of birth) field, to prevent underage users to register to the product, because of the "COPPA".

I have considered just putting in the schema field type "number" with the name age and set the minimum, but it would also be nice to be able to apply cookie, but also instead of using the number filed to use the date

If someone has a good suggestion or solution, I would be happy to hear it out. Because the easiest thing now I am thinking about is to do the limitation in the front end, but I don't think that is enough

### PR DESCRIPTION
**Pull request by harnash**
This change allows webhooks to control the registration flow by either allowing process to continue or fail with an errors propagated to the UI (mostly validation errors at this point). This is not yet complete but the general idea is here to start any discussions and resolve concerns early. 

TODO:
* tests
* documentation update

Closes #1724

## Related issue(s)
https://github.com/ory/kratos/discussions/1483

## Checklist
- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by aeneasr** [comment]
Thank you for the PR! I will review it next week!

**Comment by harnash** [comment]
> Thank you for the PR! I will review it next week!

No hurry there. I'm off next week so effectively I can resume work on this in two weeks more or less.

**Comment by aeneasr** [comment]
Could you please enable [write access for maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)? 🤔 

Thank you! 🙏

**Comment by aeneasr** [comment]
Alternatively, given that this is an org repo, you can also invite me as a collaborator to the fork

File: driver/config/.schema/config.schema.json
Comment (aeneasr) [inline driver/config/.schema/config.schema.json:152]: Let's make this explicit:

```suggestion
            "can_interrupt": {
```

File: driver/config/.schema/config.schema.json
Comment (aeneasr) [inline driver/config/.schema/config.schema.json:155]: Please also explain to which flows this applies:

```suggestion
              "description": "If enabled allows the web hook to interrupt / abort the self-service flow. It only applies to a certain flows (EXPLAIN WHICH ONES) and requires a valid response format."
```

File: schema/errors.go
Comment (aeneasr) [inline schema/errors.go:192]: Shouldn't we also be able to see the concrete errors?

File: schema/errors.go
Comment (aeneasr) [inline schema/errors.go:199]: How about:

```suggestion
func (e ValidationListError) HasErrors() bool {
```

File: ui/container/container.go
Comment (aeneasr) [inline ui/container/container.go:193]: Remove

File: ui/container/container.go
Comment (aeneasr) [inline ui/container/container.go:192]: Add a test for this

File: schema/errors.go
Comment (aeneasr) [inline schema/errors.go:187]: Add tests for this struct

File: selfservice/hook/web_hook.go
Comment (aeneasr) [inline selfservice/hook/web_hook.go:75]: I don't think we should include these. I believe they would be empty always anyways and it would not be a good idea to send e.g. hashes over the network. The risk of accidental logging by some component or other types of compromise is just too high! So I say we remove this:

```suggestion
```

File: selfservice/hook/web_hook.go
Comment (aeneasr) [inline selfservice/hook/web_hook.go:353]: So we have two options here:

1. If the response is an error code we stop execution and try to understand the error, relaying it to the UI
2. If the response contains a special field (e.g. `abort_flow`) we interrupt the flow and display an error

Generally speaking I think calling the config `can_interrupt` is misleading, because an error will always interrupt the flow. What this feature is doing is to allow us to gracefully return validation errors to the user, as opposed to showing a 500 server error.

File: selfservice/hook/web_hook.go
Comment (aeneasr) [inline selfservice/hook/web_hook.go:378]: Remove this - see comment above

```suggestion
```

File: selfservice/hook/web_hook.go
Comment (aeneasr) [inline selfservice/hook/web_hook.go:353]: This actually has a connection leak. It should be written like so and also be using a resilient client:

```go
	client := httpx.NewResilientClient(/*...*/)
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= http.StatusBadRequest {
```

You can instantiate the resilient client once in the hook struct `New` method.

File: selfservice/hook/web_hook.go
Comment (aeneasr) [inline selfservice/hook/web_hook.go:370]: Instead of reading the body into memory and then unmarshalling, it is much more effective to do:

```
if err := json.NewDecoder(resp.Body).Decode(); err != nil {
```

**Review by aeneasr — CHANGES_REQUESTED** [review]
Thank you for the PR! This is going in the right direction :)

I have merged the conflicts (as we have changed something in the hooks) which I pushed to a separate branch and made a PR against your fork: https://github.com/Wikia/kratos/pull/14

High-level this looks good. To see it in action, we will need some tests. I am in particular interested to see how the error parsing will be working in combination with the UI messages / fields.

We also need to make some decisions on how Ory Kratos detects whether a hook wants to return validation errors or not. I also think we need to rename the feature to better match what it does.

* [ ] Define a better name for the feature in the config
* [ ] Come up with a definition how a web hook could tell Ory Kratos that we have a validation error going on
* [ ] Add functional tests for web hook
* [ ] Add unit tests for the new helper structs
* [ ] Add documentation on how to return validation errors for hooks
* [ ] Add e2e test that tests this feature from start to finish
* [ ] Remove credentials from hooks

File: schema/errors.go
Comment (harnash) [reply schema/errors.go]: Fix in 75a55945

File: schema/errors.go
Comment (harnash) [reply schema/errors.go]: Fix in 97d539aa

File: selfservice/hook/web_hook.go
Comment (harnash) [reply selfservice/hook/web_hook.go]: Removed in d8511ebd

File: selfservice/hook/web_hook.go
Comment (harnash) [reply selfservice/hook/web_hook.go:361]: I was thinking about it in a similar way and I've decided to go with HTTP status codes to be somewhat in sync with what the Kratos client will receive.

Option 1. Is a bit more logical from the hook perspective that 200s will mean success and will not interrupt the flow and any other status code will one way or the other cause errors.
Option 2. Is a bit cleaner on Kratos' side since it will explicitly indicate an event that is being triggered by the web hook reponse.

I can go with either one and do not have strong preferences. I guess it is your call then @aeneasr :-)

Comment (harnash) [reply selfservice/hook/web_hook.go:364]: Done in 6b4c11081a322b71ff44231a9ead616580654442

File: selfservice/hook/web_hook.go
Comment (harnash) [reply selfservice/hook/web_hook.go]: That is neat! Thanks for the tip!  Fix in: 81106d15.

File: ui/container/container.go
Comment (harnash) [reply ui/container/container.go]: Damn. Fixed in 7177daf8

**Comment by dadrus** [comment]
When I was implementing the web hooks, my extension idea was to let the response be verified in a yet another jsonet template to which, as with building of the web hook request, a couple of objects can be passed in, like the response and the actuall flow objekt. This way the protocol of the web hook is fully open and one can achieve much more compared with the implementation in this pr. What do you think?

**Comment by harnash** [comment]
> When I was implementing the web hooks, my extension idea was to let the response be verified in a yet another jsonet template to which, as with building of the web hook request, a couple of objects can be passed in, like the response and the actuall flow objekt. This way the protocol of the web hook is fully open and one can achieve much more compared with the implementation in this pr. What do you think?

Yes I was thinking about this but I am not sure how we would properly propagate those responses up to the UI which requires some concrete structures/data to properly handle validation errors etc. (see: https://github.com/ory/kratos/blob/master/ui/container/container.go#L152). It would be great to come up with some flexible solution that would allow to validate response using Jsonnet and transform it into something Kratos can use. For the scenarion where  we have fixed data structures in Kratos and create custom payload for the webhooks Jsonnet works nicely providing the glue betwen Kratos and external services for the other way around it is a bit trickier. There must be some binding point. I will think about it more but right now I have no good solution.

**Comment by dadrus** [comment]
I think this can remain as is implemented by this PR. I mean the implementation right now expects a message with some text in the web hook response. Why not just letting a jennet template call a function on the passed context object, like e.g. `ctx.abortFlow("some error description")` ."some error description" can then come from the payload received from the Web hook or be just set in the jsonet template based on some conditions which are up to the developer of that template. When the template execution is done the remaining code will have everything which is required. Or do I overlook something?

**Comment by dadrus** [comment]
If you like I can contribute some lines of code 😉 so you can directly see if it fits.

**Comment by aeneasr** [comment]
JsonNet is not really a programming language that executes things. It's for manipulating JSON! Adding functionality like `ctx.abortFlow` will not work because the control flow is not aborted - you still need to return some JSON. I also think it is not a good idea to have this (potentially untrusted) code be able to call into the internal API of Ory Kratos (as a general pattern, not particularly for this idea).

Keep in mind that regular code is much easier to program, test, and document. JsonNet is not that great when trying to test it. The less logic we have in JsonNet, the better!

**Comment by aeneasr** [comment]
While the PR is being worked on I will mark it as a draft. That declutters our review backlog :)

Once you're done with your changes and would like someone to review them, mark the PR as ready and request a review from one of the maintainers.

Thank you!

**Comment by harnash** [comment]
OK. I had some interruptions lately and I wasn't able to devote much time to this but hopefully I can resume work on this soon.

**Comment by aeneasr** [comment]
No rush! Unmerged PRs give me a bit of anxiety so I tend to try to keep the  backlog short 😅

**Comment by aeneasr** [comment]
When this is good for review, mark it as ready!

**Comment by harnash** [comment]
I've just added support for maintaining form/flow data when hook returns error. Only tests to go now I think.

**Comment by harnash** [comment]
Sorry again for the delay. I was quite busy but I'm going to work on it this week. Will ping you when I'm done.

File: schema/errors.go
Comment (harnash) [reply schema/errors.go:260]: Done in d188b7c55

File: ui/container/container.go
Comment (harnash) [reply ui/container/container.go:203]: Done in d188b7c55

**Comment by harnash** [comment]
I think it is ready for review

**Comment by aeneasr** [comment]
@harnash sorry I was really busy these past days. I will try to take a 👀 asap

**Comment by harnash** [comment]
One question to answer is: how to handle errors when parsing responses from the blocking/interruptible web hooks? 

Currently when there is an issue when:
* response could not be unmarshaled
* could not call webhook
Kratos will return 5xx error resulting in selfservice `/error` redirection. From user perspective it might be better to return generic error (i.e. error calling webhook) and retain form and it's contents so it can be retried if needed. Right now it is not possible and user has to retype all data in the forms.

**Comment by harnash** [comment]
@aeneasr did you had time to take a look?

File: selfservice/flow/login/hook.go
Comment (aeneasr) [inline selfservice/flow/login/hook.go:72]: I think this is missing tests?

File: selfservice/flow/recovery/hook.go
Comment (aeneasr) [inline selfservice/flow/recovery/hook.go:67]: Missing tests?

File: selfservice/flow/registration/hook.go
Comment (aeneasr) [inline selfservice/flow/registration/hook.go:88]: Missing tests?

File: selfservice/flow/settings/hook.go
Comment (aeneasr) [inline selfservice/flow/settings/hook.go:99]: Missing tests?

File: selfservice/flow/verification/hook.go
Comment (aeneasr) [inline selfservice/flow/verification/hook.go:67]: Missing tests?

File: ui/container/container.go
Comment (aeneasr) [inline ui/container/container.go:90]:
```suggestion
// NewFromStruct creates a UI Container based on serialized contents of the provided struct.
```

File: selfservice/flow/registration/hook.go
Comment (aeneasr) [inline selfservice/flow/registration/hook.go:96]: Can you please extract this logic into a common helper function that converts from credential types to groups and vice versa? And then use this instead of having the correct group conversion applied in different files?

**Comment by aeneasr** [comment]
@abador I have some changes locally which I would like to push to your branch. You invited me to the repo but I did not see it and the invitation has expired. Could you please invite me again? Sorry for the trouble :(

**Comment by aeneasr** [comment]
Closed on accident

File: selfservice/hook/web_hook.go
Comment (aeneasr) [reply selfservice/hook/web_hook.go]: The connection leak is still there, reopening this. I have a local commit which fixes this.

**Comment by harnash** [comment]
@aeneasr I've just re-added you with `write` access

**Comment by aeneasr** [comment]
Thank you! I pushed my changes!

**Review by aeneasr — COMMENTED** [review]
Thank you, this looks great! I think we need a couple more things before we can merge this into master :)

1. [ ] It would be really good to have E2E tests for every flow that ensure that the web hook failure indeed leads to a UI message. Only that way can we be certain that it works as expected.
2. [ ] The documentation (as noted in the opening message) should be updated to reflect how hooks can intercept registration.
3. [ ] The new methods added to the hooks (e.g. `handleLoginError`, `handleRegistrationError`) are more or less duplicated and all do the same thing. Can we maybe extract them into a common helper function which is tested, and reuse the code?

Something else I noticed during code review is that you don't like the `json:""` tags :) For Ory's APIs we always use lower case and underscore naming for JSON. So `InstancePtr` becomes `instance_pointer` :) I updated the place accordingly wherever I found this to be missing!

**Comment by aeneasr** [comment]
By the way, I am keen on getting this merged :) So will try to be as responsive as possible!

**Comment by harnash** [comment]
Thanks. I will have time to work on this soon. I'll read through all comments and try to address them quickly!

**Comment by harnash** [comment]
I've added the docs but I'm waiting for a fix to:
```json
Unable to build website for locale "en".
Error: Server responded to https://raw.githubusercontent.com/ory/docs/master/docs/docs/ecosystem/bugbounty.md with status code 404:
404: Not Found
``` which seems to be caused by https://github.com/ory/docs/pull/540

**Comment by vinckr** [comment]
Fixed now @harnash, my bad!

**Comment by lanphan** [comment]
Hello @aeneasr and @harnash ,

There is 1 special case here with registration flow: in case registration flow is configured with important web-hooks (such as hook to continue some business logic after creating new identity in kratos), is it possible to use this new feature, interuptable web-hook, to configure failed flow with some additional actions (such as, delete new created identity because web-hook failed)?

(In fact, I see that there is PostRegistrationPrePersistHooks in code, but I don't know how to configure it, seems it's not available)

**Comment by harnash** [comment]
@lanphan Right. So you could run some additional logic before identity is created (but not yet stored in the database) via `PostRegistrationPrePersistHook` and do complex business logic. That is exactly what we are doing right now. It is a bit tricky since there are timeouts (so the hook can't run for too long) but in most cases it should be OK.

**Comment by lanphan** [comment]
thank @harnash ,
looking forward to your PR :)

File: selfservice/flow/login/hook_test.go
Comment (abador) [inline selfservice/flow/login/hook_test.go:52]: shouldn't tests cover other types? https://github.com/Wikia/kratos/blob/upstream_blocking_hooks/identity/credentials.go#L46

**Comment by harnash** [comment]
Alright! I think it is ready for re-review!

File: selfservice/flow/registration/hook.go
Comment (harnash) [reply selfservice/flow/registration/hook.go:120]: Will do!

**Comment by aeneasr** [comment]
While the PR is being worked on I will mark it as a draft. That declutters our review backlog :)

Once you're done with your changes and would like someone to review them, mark the PR as ready and request a review from one of the maintainers.

Thank you!

**Comment by harnash** [comment]
I should be done with the fixes next week. Right now I have to focus on other tasks. Sorry for the delay.

**Comment by aeneasr** [comment]
No problem!

**Comment by harnash** [comment]
Fix in 6a08bef8d8b4394fd834b70da6d7fc91df241c1d

**Comment by harnash** [comment]
I believe I've addressed all comments.

**Review by hperl — CHANGES_REQUESTED** [review]
Hi, thank you for your changes. I have taken over the review from @aeneasr and left some comments so that we can get this merged soon :muscle: . I am still somewhat new to the code base, so please forgive me if some comments are not applicable :wink: .

**Comment by harnash** [comment]
@hperl thank you. I will be looking and addressing those comments today and tomorrow.

File: identity/credentials.go
Comment (harnash) [reply identity/credentials.go:58]: Fix in f7ef178ea

File: selfservice/flow/login/hook_test.go
Comment (harnash) [reply selfservice/flow/login/hook_test.go:33]: Fix in f7ef178ea

File: schema/errors.go
Comment (harnash) [reply schema/errors.go:251]: Right. This one is a bit special since it is wrapped by `ValidationListError`. I've added `errors.Wrap()` to `NewValidationListError()` but I don't think it will be good to wrap those additionally.

File: selfservice/flow/login/handler.go
Comment (harnash) [reply selfservice/flow/login/handler.go:175]: I was looking at `selfservice/flow/registration/handler.go` it is a bit similar but login flow might behave a bit differently than registration. I'll add a check that will prevent exception but I'm not sure what to choose there. Logic is a bit different and we do not stop after first successful strategy completed so I would not like to change this behaviour unless I'm pretty sure it is the right way to do it. We could use `node.DefaultGroup` but that would probably be problematic in some cases.

**Comment by hperl** [comment]
Thanks, let me know when I can take a look :).

File: selfservice/flow/login/handler.go
Comment (harnash) [reply selfservice/flow/login/handler.go:613]: I think this one is in sync with registration. We use the first one that succeeded (last iteration). If there were errors than we deal with that in the loop above if none succeeded (or none are defined) then we do not have a valid identity in `i` and the process will halt before reaching this point.

File: selfservice/hook/web_hook.go
Comment (harnash) [reply selfservice/hook/web_hook.go:178]: Partially out of habit :-) But this is an implementation assumption and caller might not be aware of this  and can cause issues when someone modifies `parseWebhookResponse()` and add a case where `nil` will be returned. Usually when function returns an `error` one might expect it to return `nil` in some cases (default Go contract) and keeping this logic would violate "least surprise" rule ;-)

File: selfservice/hook/web_hook.go
Comment (harnash) [reply selfservice/hook/web_hook.go:230]: Umm. So there is an case when hook is set to be "blocking" type. There is an contract/assumption that there webhook will either:
* succeed and flow will continue as normal (http status < 400)
* will process the request and decide to abort the flow (http status >= 400) and now we might have situation:
  * webhook returned response that Kratos could decode and could understand and propagates error to the UI
  * webhook returned outdated/corrupted response and Kratos could not parse it correctly
  * request was handled/interrupted somewhere between Kratos and webhook and response is also unrecognizable by Kratos (basically the same situation as above)
  * webhook returned an error but there are no errors inside a container (something wrong happened inside and probably needs attention, it looks like webhook decided to block the flow but couldn't add proper error) <- this is this case

It is probably not so convenient but we need to take into account that request may fail and Kratos should react somehow. It is just an extra check to make sure that webhook returns valid data so we don't break flow without specifying a reason for the user to react.

**Comment by aeneasr** [comment]
To resolve the CI I think we need to resolve the merge conflicts :)

**Comment by aeneasr** [comment]
There's still one failing test :) https://github.com/ory/kratos/runs/6671177484?check_suite_focus=true#step:16:20427

**Comment by harnash** [comment]
@hperl I think I'm done addressing your comments. Let me know if there  is anything else.

**Comment by hperl** [comment]
Thanks for the ping, I'll take a look tomorrow!

File: selfservice/hook/web_hook.go
Comment (hperl) [reply selfservice/hook/web_hook.go:200]: One more thing, we only read from `errChan` once in https://github.com/ory/kratos/pull/1585/files#diff-6dbaf3f7f3f6f427e72de26f22beb17b0c56379f702ae5be81516c38765f8d13R220, but are writing to it in lines 201 and 204. The second send (in 204) will still succeed, since the channel is buffered, but it is still cleaner to return after the first send in 201, since that is what will really happen, right?

**Review by hperl — DISMISSED** [review]

File: embedx/config.schema.json
Comment (aeneasr) [inline embedx/config.schema.json:211]: What happens if we set this value and `response.ignore` both to true? Should they be mutually exclusive?

**Comment by aeneasr** [comment]
All in all this looks very, very, very good :) I will need to try it out end-to-end, but I don't think that there will be big topics outside of: https://github.com/ory/kratos/pull/1585#discussion_r891441712 :) :)

File: embedx/config.schema.json
Comment (harnash) [reply embedx/config.schema.json:211]: Will take a look.

Comment (harnash) [reply embedx/config.schema.json:211]: Not an JSON schema expert but I think I got this working: 97514c298

**Comment by harnash** [comment]
I think I've addressed all comments @aeneasr + @hperl . We are going to try migrate to v0.10.x in few days and if you say it is in mergable state I will try to incorporate this into our branch and test it on DEV env to see if there are some unexpected issues.

**Comment by aeneasr** [comment]
> I think I've addressed all comments @aeneasr + @hperl . We are going to try migrate to v0.10.x in few days and if you say it is in mergable state I will try to incorporate this into our branch and test it on DEV env to see if there are some unexpected issues.

Do you have an update on this? :)

**Comment by harnash** [comment]
> > I think I've addressed all comments @aeneasr + @hperl . We are going to try migrate to v0.10.x in few days and if you say it is in mergable state I will try to incorporate this into our branch and test it on DEV env to see if there are some unexpected issues.
> 
> Do you have an update on this? :)

Sadly migration to v0.10 is being postponed a bit. Will let you know as soon as I know more. Currently we have this in backlog and planned for next sprint (~ two weeks).

**Comment by aeneasr** [comment]
Ok, no problem :) I think the code looks good so I would be fine merging it after I do one final round of review

**Review by aeneasr — APPROVED** [review]
It's perfect. Thank you!

I've made two minor changes: I removed the `message` string and renamed `detailed_messages` to `messages`. The reason being that `message` did not actually have an effect and it was a bit confusing why we need it.

The second thing was that there was a connection leak in the webhook handler as `resp.Body.Close()` wasn't being called. This is fixed now :)

I've updated the docs also

Overall, great achievement!

**Comment by harnash** [comment]
Thanks!
We are still working on upgrade to v0.10.1 but I think we could test this build soon enough.

**Pull request merged**
